### PR TITLE
Basic codefreeze

### PR DESCRIPTION
**Pull request by kateEcobb**
Implements codefreeze action, with an allowlist of paths that can override the code freeze. You can run tests via `cd codefreeze && yarn test`.

next up will be creating a reusable workflow and skipping the action entirely (so we don't add to our github actions usage if a code freeze isn't in effect) but I wanted to get this out for review. You can see the effect of the action in the core-frontend repo: https://github.com/goodwatercap/core-frontend/pull/475

### CODE REVIEW

File: codefreeze/src/main.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+      // Ensure that the head commit is ahead of the base commit.
+      if (response.data.status !== "ahead") {
+        core.setFailed(
+          `The head commit for this ${context.eventName} event is not ahead of the base commit.`
+        );
+      }
+
+      const files = response.data.files;
+      const parsedPaths = allowedPaths.split(/\r|\n/);
```
Comment (kyoung) [inline codefreeze/src/main.ts:80]: Might be worth commenting in the top description what encoding you're expecting these to be in (eg a yaml list, a comma sep list, etc)

File: codefreeze/action.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+# Define your inputs here.
+inputs:
+  codefreeze-begin:
+    description: "The date when the code freeze begins"
+    required: true
+  codefreeze-end:
+    description: "The date when the code freeze ends"
+    required: true
+  allowed-paths:
+    description: "The paths that are allowed to be modified during the code freeze"
```
Comment (kyoung) [inline codefreeze/action.yml:14]: See note below about required format

File: codefreeze/src/main.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        core.info("No staged files. Exiting.");
+        return;
+      }
+
+      for (const file of files) {
+        const filename = file.filename;
+        core.debug(`Checking file: ${filename}`);
+
+        const isAllowed = some(parsedPaths, (path) => {
+          return filename.startsWith(path);
```
Comment (kyoung) [inline codefreeze/src/main.ts:93]: Some part of me suspects that `startsWith` is going to have edge cases as a basis for our "allowed exceptions" file comparator, but the cold medicine is blocking examples 😬 

I feel like allowed paths should properly be folders (check if child, must end in `/`), specific files (check for exact match), and search strings (what we've got going here, but ending in `*` or something). However I think this 💯 for a first go.

**Review by kyoung — APPROVED** [review]

File: codefreeze/src/main.ts (hunk 3 above)
Comment (kateEcobb) [reply codefreeze/src/main.ts:93]: yeah for sure, Im going to clean up things in another pr, i just wanted to get the first iteration out there!

**Pull request merged**